pages/posts/ocaml-writer.mdx

```diff
@@ -0,0 +1,234 @@
+
+Feel free to skip the background section if you don't care how this problem came up in a real project.
+
+## Background
+
+The [Coda protocol](https://codaprotocol.com) uses [zk-SNARKs](TODO) to compress the blockchain down to a constant size. These SNARKs are proof objects certifying that some computation has been run correctly. Some nodes in the network create these SNARKs and others verify them. To create a SNARK one needs to have a specific large _proving key_[^1], and to verify a SNARK one needs to a specific small _verification key_[^2]. These proving and verification keys are indexed by the SNARK logic itself. In other words, if any computation changes than the proving and verifying keys need to change. In other other words, whenever some code is changed in the SNARK circuit during the development of Coda, both proving keys and verification keys must be regenerated. Key generation also involves some randomness.
+
+We want developers to be productive and so we want proving and verification key generation to happen transparently at build time.
+
+Naively, one can just always regenerate all proving and verification keys every build, but this has some drawbacks: (1) As randomness is involved during key generation, keys created from different builds will be incompatible. For example, you wouldn't be able to connect to a live network that was built on CI[^3] with some local branch of code, even though when you haven't touched any SNARK logic that would invalidate the keys. (2) Also, key generation is slow so we'd like to skip that step if possible.
+
+A nice solution to this would be to remove the randomness from key generation under certain conditions. This may be done by adding some debug branch through the code that generates . Unfortunately, it turned out to be quite a bit of work to get rid of all the randomness in the key generation logic. It's enough work that we decided to punt on it temporarily.
+
+Another solution is to introduce a series of layered caches[^4] that would be placed in front of key-generation in the key-loading process. To load keys in Coda, the following process is followed:
+
+1. Try to load the keys from a manual override path
+1. Try to load the keys from the normal installation path
+1. Try to load the keys from an S3 installation path
+1. Try to load the keys from S3
+1. Try to load the keys from the auto-generation path
+1. Auto-generate the keys
+1. Store the keys in the auto-generation path (only if the keys were auto-generated)
+1. Store the keys in S3 (only if the keys were auto-generated)
+
+To further complicate things, we need to propagate the knowledge that we want to store keys in S3 through several routines across a variety of different modules, and signal later if any one load caused some sort of regeneration. This is a combination of this part of the code being perhaps in need of a full-rewrite as it's been incrementally changed for two years and other unimportant reasons[^5].
```

Comment:
I wonder if this can be simplified to "To further complicate things, we need to track the outcome of the key generation process and propagate it to various interested observers for legacy reasons."